### PR TITLE
Update exp package

### PR DESCRIPTION
**Pull request by gouthamve**
imo, this is worse in term or readability :(

Fixes https://github.com/prometheus/prometheus/issues/12649

### CODE REVIEW

**Review by jesusvazquez — APPROVED** [review]
I dont mind the readability changes I think we'll get use to them.

**Comment by gouthamve** [comment]
So the test failure is interesting. Casting a float to an int leads to rounding to `0` which means equivalence in the compare function.

How to handle it in a simple manner is eluding me. Thoughts anyone? cc @bboreham @roidelapluie  

In `promql/quantile.go`:

```diff
-	slices.SortFunc(buckets, func(a, b bucket) bool {
-		return a.upperBound < b.upperBound
+	slices.SortFunc(buckets, func(a, b bucket) int {
+		return int(a.upperBound - b.upperBound)
```


https://play.golang.com/p/M03btR83odb

**Comment by gouthamve** [comment]
There is https://pkg.go.dev/cmp#Compare but it's not released yet and bumps the requirement to Golang 1.21

**Comment by bboreham** [comment]
I think we should benchmark this; the extra work required to produce ternary results is relatively high.  Also `strings.Compare` is deliberately not optimized, for reasons that are now invalid.

If it turns out this hurts performance, noting that the implementation of `SortFunc` only ever uses the function as `cmp(...) < 0`, we could cheat and only implement that side.

Maybe Go 1.21 has other improvements that regain the lost performance. Let's measure, anyway.

Regarding the float comparison, it seems the Go team went for the complicated version: https://github.com/golang/exp/blob/d63ba01acd4b5892385c4390b334ea2c89f27eb7/slices/cmp.go#L30-L44

File: model/labels/labels.go [hunk 1]
```diff
@@ -46,9 +47,11 @@ type Label struct {
 // instantiation.
 type Labels []Label
 
-func (ls Labels) Len() int           { return len(ls) }
-func (ls Labels) Swap(i, j int)      { ls[i], ls[j] = ls[j], ls[i] }
-func (ls Labels) Less(i, j int) bool { return ls[i].Name < ls[j].Name }
+func (ls Labels) Len() int      { return len(ls) }
+func (ls Labels) Swap(i, j int) { ls[i], ls[j] = ls[j], ls[i] }
+func (ls Labels) Less(i, j int) int {
```
Comment (bboreham) [inline model/labels/labels.go:52]: This one strikes me as wrong, as the interface is intended for `sort.Sort` which hasn't changed.

**Comment by bboreham** [comment]
For at least one benchmark, which is dominated by the sorting time, this change made no difference:

```
cpu: Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz
          │  before.txt  │          after.txt           │
          │    sec/op    │   sec/op     vs base         │
Builder-4   784.0n ± 30%   781.8n ± 4%  ~ (p=1.000 n=6)

          │ before.txt │           after.txt           │
          │    B/op    │    B/op     vs base           │
Builder-4   288.0 ± 0%   288.0 ± 0%  ~ (p=1.000 n=6) ¹
¹ all samples are equal

          │ before.txt │           after.txt           │
          │ allocs/op  │ allocs/op   vs base           │
Builder-4   1.000 ± 0%   1.000 ± 0%  ~ (p=1.000 n=6) ¹
¹ all samples are equal
```

File: cmd/promtool/tsdb.go [hunk 2]
```diff
@@ -446,7 +446,7 @@ func analyzeBlock(path, blockID string, limit int, runExtended bool) error {
 	postingInfos := []postingInfo{}
 
 	printInfo := func(postingInfos []postingInfo) {
-		slices.SortFunc(postingInfos, func(a, b postingInfo) bool { return a.metric > b.metric })
+		slices.SortFunc(postingInfos, func(a, b postingInfo) int { return int(b.metric - a.metric) })
```
Comment (bboreham) [inline cmd/promtool/tsdb.go:449]: `b.metric` is unsigned, so you're relying on overflow. Either cite why this works, or cast to `int` before subtracting.

File: promql/quantile.go [hunk 3]
```diff
@@ -81,8 +81,14 @@ func bucketQuantile(q float64, buckets buckets) float64 {
...
-	slices.SortFunc(buckets, func(a, b bucket) bool {
-		return a.upperBound < b.upperBound
+	slices.SortFunc(buckets, func(a, b bucket) int {
+		if a.upperBound < b.upperBound {
+			return -1
+		}
+		if a.upperBound > b.upperBound {
+			return +1
+		}
+		return 0
```
Comment (bboreham) [inline promql/quantile.go:91]: Perhaps worth commenting that we don't expect NaN?

File: rules/manager.go [hunk 4]
```diff
@@ -1203,11 +1206,11 @@ func (m *Manager) RuleGroups() []*Group {
 		rgs = append(rgs, g)
 	}
 
-	slices.SortFunc(rgs, func(a, b *Group) bool {
+	slices.SortFunc(rgs, func(a, b *Group) int {
 		if a.file != b.file {
-			return a.file < b.file
+			return strings.Compare(a.file, b.file)
```
Comment (bboreham) [inline rules/manager.go:1211]: Use the zero return from `strings.Compare` instead of checking equality first.

File: tsdb/index/postings.go [hunk 5]
```diff
@@ -107,11 +108,11 @@ func (p *MemPostings) SortedKeys() []labels.Label {
 	}
 	p.mtx.RUnlock()
 
-	slices.SortFunc(keys, func(a, b labels.Label) bool {
+	slices.SortFunc(keys, func(a, b labels.Label) int {
 		if a.Name != b.Name {
-			return a.Name < b.Name
+			return strings.Compare(a.Name, b.Name)
```
Comment (bboreham) [inline tsdb/index/postings.go:113]: Use the zero return from `strings.Compare` instead of checking equality first.

**Review by bboreham — COMMENTED** [review]
Some things I think should change.

**Comment by gouthamve** [comment]
@bboreham Good catches! Could you take another look?

**Review by metalmatze — APPROVED** [review]
Looking great! 
Anything left to do to get this merged?

**Review by bboreham — COMMENTED** [review]
Please check 1 item.

File: tsdb/index/postings.go [hunk 6]
```diff
@@ -107,11 +108,14 @@ func (p *MemPostings) SortedKeys() []labels.Label {
 	}
 	p.mtx.RUnlock()
 
-	slices.SortFunc(keys, func(a, b labels.Label) bool {
-		if a.Name != b.Name {
-			return a.Name < b.Name
+	slices.SortFunc(keys, func(a, b labels.Label) int {
+		nameCompare := strings.Compare(a.Name, b.Name)
+		// If names are the same, compare values.
+		if nameCompare == 0 {
```
[the remark this answers is not in the document]
Comment (gouthamve) [reply tsdb/index/postings.go]: You are right. Fixed.

**Pull request merged**